internal/metrics/sources/prometheus/convert.go

```diff
@@ -0,0 +1,238 @@
+	case int64:
+		return toInstrumentDataPoint[int64](input, result)
+	case float64:
+		return toInstrumentDataPoint[float64](input, result)
+	default:
+		return metricdata.Metrics{}, fmt.Errorf(
+			"could not convert data entry of value type [%T] to OTel metric", input.Values[0].Value,
+		)
+	}
+}
+
+func toInstrumentDataPoint[N float64 | int64](
+	input model.DataEntry, base metricdata.Metrics,
+) (metricdata.Metrics, error) {
+	switch input.Type {
+	case model.Counter:
+		base.Data = toCounter[N](input)
+	case model.Gauge:
+		base.Data = toGauge[N](input)
+	case model.Histogram:
+		base.Data = toHistogramDataPoint[N](input)
+	case model.Summary:
+		slog.Debug("unhandled metrics conversion of type 'summary': not supported yet")
+		return metricdata.Metrics{}, nil
+	case model.Unknown:
+		fallthrough
+	default:
+		slog.Debug("unhandled metrics conversion", "type", input.Type)
+		return metricdata.Metrics{}, fmt.Errorf("unhandled metrics conversion of type: %s", input.Type)
+	}
+
+	return base, nil
+}
+
+func toCounter[N float64 | int64](input model.DataEntry) *metricdata.Sum[N] {
+	dataPoints := make([]metricdata.DataPoint[N], 0)
+
+	for _, point := range input.Values {
+		metricAttributes := make([]attribute.KeyValue, 0)
+		for labelKey, labelValue := range point.Labels {
+			metricAttributes = append(metricAttributes, attribute.KeyValue{
+				Key: attribute.Key(labelKey), Value: attribute.StringValue(labelValue),
+			})
+		}
+
+		value, ok := point.Value.(N)
+		if !ok {
+			// Do not return error for malformed data but instead ignore the data point.
+			slog.Debug("could not cast metric data point value to correct type: discarding data point", "value", value)
+			continue
+		}
+
+		dataPoints = append(dataPoints, metricdata.DataPoint[N]{
+			Attributes: attribute.NewSet(metricAttributes...),
+			Time:       time.Now(),
+			Value:      value,
+		})
+	}
+
+	return &metricdata.Sum[N]{
+		DataPoints:  dataPoints,
+		Temporality: metricdata.DeltaTemporality,
+		IsMonotonic: true,
+	}
+}
+
+func toGauge[N float64 | int64](input model.DataEntry) *metricdata.Gauge[N] {
+	dataPoints := make([]metricdata.DataPoint[N], 0)
+
+	for _, point := range input.Values {
+		metricAttributes := make([]attribute.KeyValue, 0)
+		for labelKey, labelValue := range point.Labels {
+			metricAttributes = append(metricAttributes, attribute.KeyValue{
+				Key: attribute.Key(labelKey), Value: attribute.StringValue(labelValue),
+			})
+		}
+
+		value, ok := point.Value.(N)
+		if !ok {
+			// Do not return error for malformed data but instead ignore the data point.
+			slog.Debug("could not cast metric data point value to correct type: discarding data point", "value", value)
+			continue
+		}
+
+		dataPoints = append(dataPoints, metricdata.DataPoint[N]{
+			Attributes: attribute.NewSet(metricAttributes...),
+			Time:       time.Now(),
+			Value:      value,
+		})
+	}
+
+	return &metricdata.Gauge[N]{
+		DataPoints: dataPoints,
+	}
+}
+
+const (
```

Comment:
prefer consts to be used near the top of the file

Reply:
https://github.com/nginx/agent/pull/544#discussion_r1483071548